unified_planning/model/ma_problem.py

```diff
@@ -0,0 +1,683 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""This module defines the MultiAgentProblem class."""
+
+import sys
+import copy
+from unified_planning.shortcuts import *
+import unified_planning.model.operators as op
+from unified_planning.model.types import domain_size, domain_item
+from unified_planning.environment import get_env, Environment
+from unified_planning.model.walkers import OperatorsExtractor
+from unified_planning.model.fluent import Fluent
+from unified_planning.model.action import *
+from fractions import Fraction
+from typing import List, Dict, Set, Union, Optional
+from unified_planning.exceptions import (
+    UPProblemDefinitionError,
+    UPTypeError,
+    UPValueError,
+    UPExpressionDefinitionError,
+)
+from unified_planning.model.agent import Agent
+from unified_planning.io.pddl_writer import PDDLWriter
+from unified_planning.model.environment_ma import Environment_ma
+import collections
+from typing import List, Union, cast
+from unified_planning.io.pddl_writer_ma import PDDLWriter_MA
+from unified_planning.model.walkers.substituter import Substituter
+import re
+import os
+from typing import IO
+from io import StringIO
+
+from unified_planning.model.abstract_problem import AbstractProblem
+from unified_planning.model.mixins import (
+    ActionsSetMixin,
+    FluentsSetMixin,
+    ObjectsSetMixin,
+    UserTypesSetMixin,
+    AgentsSetMixin,
+)
+
+
+class MultiAgentProblem(
+    # Problem,
+    AbstractProblem,
+    UserTypesSetMixin,
+    FluentsSetMixin,
+    ActionsSetMixin,
+    ObjectsSetMixin,
+    AgentsSetMixin,
+):
+    """Represents a planning MultiAgentProblem."""
+
+    def __init__(
+        self,
+        name: str = None,
+        env: "up.environment.Environment" = None,
+        env_ma: Union["up.model.environment_ma.Environment_ma", None] = None,
+    ):
+        AbstractProblem.__init__(self, name, env)
+        UserTypesSetMixin.__init__(self, self.has_name)
+        FluentsSetMixin.__init__(self, self.env, self._add_user_type, self.has_name)
+        ActionsSetMixin.__init__(self, self.env, self._add_user_type, self.has_name)
+        ObjectsSetMixin.__init__(self, self.env, self._add_user_type, self.has_name)
+        AgentsSetMixin.__init__(self, self.env, self._has_name_method)
+
+        self._env_ma: Union["up.model.environment_ma.Environment_ma", None] = env_ma
+        self._initial_value: Dict["up.model.fnode.FNode", "up.model.fnode.FNode"] = {}
+        self._goals: List["up.model.fnode.FNode"] = list()
+        self._shared_data_list: List["up.model.fluent.Fluent"] = []
+        self._new_fluents: Dict["str", "up.model.fluent.Fluent"] = {}
+        self._new_objects: Dict["str", "up.model.object.Object"] = {}
+        self._shared_data: List["up.model.fluent.Fluent"] = []
+        self._flu_functions: List["up.model.fluent.Fluent"] = []
```

Comment:
_new_fluents and _new_objects will be moved to transformers (currently, needed by compile)
_shared_data_list will be dropped (it's a leftover from previous versions)
_shared_data stores shared fluents that are specific to and needed by MAProblem (will remain)
_flu_functions stores object fluents (used in, e.g., FMAP). For this, we need a specific class, ObjectFluen (derived from Fluent?), in the UP. Once available,  _flu_functions will not be needed anymore (as agent will store the object fluents/functional fluents).

Reply:
Not sure to understand what you mean by "_shared_data stores shared fluents that are specific to and needed by MAProblem" Are these Fluents a subset of Agent fluents? Or are "global" fluents visible to all agents?

For the _flu_functions filed, note that in the UP you can already have a fluent of object type:
```python
Location = UserType("Location")
Robot = UserType("Robot")
position = Fluent("position", Location, robot=Robot) # a fluent called position of type Location with a parameter of type Robot 
```
You can discriminate the fluents by simply cheching their type:
```python
position.type.is_user_type() # returns True
```

All the other fields, if needed by the compilation, can go to the dedicated compilation class in a subsequent PR. Please remove them for now.

Reply:
shared_data are fluents that the other agents can see. Which fluents are shared depends on the specific problem, not on the agent. Thus, this structure needs to be in MAProblem.

For object fluents, we will use the suggested function and remove flu_functions.

Reply:
It is not clear to me which fluents are these. Are these fluents defined in the `MAEnvironment` or in a `Agent`? 

When you say "other agents", other than what?

Reply:
They are agent fluents which are visible to the other agents in the MAProblem.

Reply:
My understanding was that we agreed on the assumption of full observability.

Reply:
Actually, there is no point in avoiding them (in fact, FMAP can deal with shared fluents). If one needs all fluents to be public, it is enough to declare them as shared. Of course, in presence of private fluents, we need to check that the targeted planner can deal with them (and if, not, return a "not supported" error). Are we missing anything here?

Reply:
You cannot express that with only a list of fluents, because the fluent does not know which agent it belongs to.

It should be something like:
`self._shared_fluents: Dict[Agent, List[Fluent]] = []`

I would also rename `shared_data` to `shared_fluents`.

If you want to support this feature already in this PR you should also extend the ProblemKind to represent the presence of private fluents and update the `kind` method to consider it.